cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,282 @@
+{
+  constexpr int block_size = 128;
+
+  auto const grid_size = (size + block_size - 1) / block_size;
+  auto d_column_info   = rmm::device_scalar<cudf::io::column_type_histogram>(stream);
+  CUDF_CUDA_TRY(cudaMemsetAsync(
+    d_column_info.data(), 0, sizeof(cudf::io::column_type_histogram), stream.value()));
+
+  infer_column_type_kernel<<<grid_size, block_size, 0, stream.value()>>>(
+    options, data, column_strings_begin, size, d_column_info.data());
+
+  return d_column_info.value(stream);
+}
+
+/**
+ * @brief Infers data type for a given JSON string input `data`.
+ *
+ * @throw cudf::logic_error if input size is 0
+ * @throw cudf::logic_error if data type inference failed
+ *
+ * @tparam ColumnStringIter Iterator type whose `value_type` is convertible to
+ * `thrust::tuple<device_span, string_view>`
+ *
+ * @param options View of inference options
+ * @param data JSON string input
+ * @param column_strings_begin The begining of an offset-length tuple sequence
+ * @param omission_null_count Number of omitted nulls
+ * @param size Size of the string input
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return The inferred data type
+ */
+template <typename ColumnStringIter>
+cudf::data_type infer_data_type(json_parse_options_view const& options,
+                                device_span<char const> data,
+                                ColumnStringIter column_strings_begin,
+                                cudf::size_type omission_null_count,
+                                std::size_t const size,
+                                rmm::cuda_stream_view stream)
+{
+  CUDF_EXPECTS(size != 0, "No data available for data type inference.\n");
+
+  auto const h_column_info = infer_column_type(options, data, column_strings_begin, size, stream);
+
+  auto get_type_id = [&](auto const& cinfo) {
+    auto int_count_total =
+      cinfo.big_int_count + cinfo.negative_small_int_count + cinfo.positive_small_int_count;
+    if ((cinfo.null_count + omission_null_count) == static_cast<cudf::size_type>(size)) {
+      // Entire column is NULL; allocate the smallest amount of memory
+      return type_id::INT8;
+    } else if (cinfo.string_count > 0) {
+      return type_id::STRING;
+    } else if (cinfo.datetime_count > 0) {
+      return type_id::TIMESTAMP_MILLISECONDS;
+    } else if (cinfo.float_count > 0 ||
+               (int_count_total > 0 && (cinfo.null_count + omission_null_count) > 0)) {
+      return type_id::FLOAT64;
+    } else if (cinfo.big_int_count == 0 && int_count_total != 0) {
+      return type_id::INT64;
+    } else if (cinfo.big_int_count != 0 && cinfo.negative_small_int_count != 0) {
+      return type_id::STRING;
+    } else if (cinfo.big_int_count != 0) {
+      return type_id::UINT64;
+    } else if (cinfo.bool_count > 0) {
+      return type_id::BOOL8;
+    } else {
+      CUDF_FAIL("Data type inference failed.\n");
+    }
```

Comment:
```suggestion
    }
    
    CUDF_FAIL("Data type inference failed.\n");
```

Reply:
@elstehle should we return a string column if input is a bunch of unquoted strings?